[SPARK-15480][UI][Streaming]show missed InputInfo in streaming UI

mwws opened the pull request:
## What changes were proposed in this pull request?

It's a bug in Streaming UI. If BatchDuration is changed by window operation, InputInfo on time with no output ops will not be shown in streaming UI.

Here is a simple example to reproduce:

``` scala
val inputDStream = KafkaUtils.createDirectStream[String, String, StringDecoder, StringDecoder](ssc, kafkaParam, topic)
val matchedDStream = inputDStream.window(batchInteval*2, batchInteval*2)
matchedDStream.foreachRDD(rdd => {
    val count = rdd.count()
    println(s"2 count number: ${count}")
})
```
## How was this patch tested?

manually tested

I manually input 6 records from Kafka. According to output, there are indeed 6 records processed
![output](https://cloud.githubusercontent.com/assets/13216322/15495140/e1f63e10-21c0-11e6-8156-f0e9694459c2.png)

but according to web UI,  input records show 0.
![original](https://cloud.githubusercontent.com/assets/13216322/15495079/82de34a0-21c0-11e6-9422-f922dff98da6.png)

**And here is the Screenshot after my change:**
![newresult](https://cloud.githubusercontent.com/assets/13216322/15702063/6bcd7aa4-2810-11e6-881e-cffed9b84acc.png)

wei-mao-intel committed:
show missed input info in streaming info page

SparkQA commented:
**[Test build #59124 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59124/consoleFull)** for PR 13259 at commit [`83351b8`](https://github.com/apache/spark/commit/83351b8f8dcb7767700157a17e513149740ab6d1).

SparkQA commented:
**[Test build #59124 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59124/consoleFull)** for PR 13259 at commit [`83351b8`](https://github.com/apache/spark/commit/83351b8f8dcb7767700157a17e513149740ab6d1).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59124/
Test PASSed.

mwws commented:
@chenghao-intel @carsonwang

chenghao-intel commented:
cc @zsxwing

jerryshao commented on streaming/src/main/scala/org/apache/spark/streaming/scheduler/JobScheduler.scala:
```diff
@@ -193,6 +197,14 @@ class JobScheduler(val ssc: StreamingContext) extends Logging {
+      inputInfoMissedTimes.foreach (time => {
+        val streamIdToInputInfos = inputInfoTracker.getInfo(time)
+        val fakeJobSet = JobSet(time, Seq(), streamIdToInputInfos)
+        listenerBus.post(StreamingListenerBatchCompleted(fakeJobSet.toBatchInfo))
```
Will this change the semantics if posing fake `StreamingListenerBatchCompleted`, if user's code rely on this, will this break their assumptions?

mwws replied:
Good point, it would be a breaking change in that case. Give the fact that some information is indeed missing now, we need either send more events or add additional fields in current event. The later might be better and more correct in semantics, although it can still break current implementation of user's custom listener. And I noticed that listener interface is annotated as DeveloperAPI, so it's expected that API might be changed. How do you think?

zsxwing commented:
I suggest we change `InputInfoTracker.getInfo(batchTime: Time)` to  `InputInfoTracker.getInfo(batchTime: Time, batchDuration: Long)` and let InputInfoTracker return the aggregation input infos between `batchTime - batchDuration` and `batchTime`.

mwws commented:
@zsxwing thank you for the suggestion, But I have two concerns: 
1. There could be multiple **output ops** on one BatchTime. And these ops could have different batchDuration. Which one should I use?
2. The _input rate graph_ is generated by the _input size_ column of batch table. If we do the aggregation, the meaning of _input rate graph_ will be changed from "number of input events from source" to "number of processed events".

zsxwing commented:
> @zsxwing thank you for the suggestion, But I have two concerns: 
> 1. There could be multiple output ops on one BatchTime. And these ops could have different batchDuration. Which one should I use?
> 2. The input rate graph is generated by the input size column of batch table. If we do the aggregation, the meaning of input rate graph will be changed from "number of input events from source" to "number of processed events".

How about adding a field `lastBatchTime` and storing the last submitted batch? Then just use the aggregated input infos between the last batch and the current batch? We need to aggregate them since we only have batch infos that have jobs.

SparkQA commented:
**[Test build #59730 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59730/consoleFull)** for PR 13259 at commit [`a3d05e0`](https://github.com/apache/spark/commit/a3d05e08c9889bc7a2d4df110430b2889858943f).

wei-mao-intel committed:
revert perivous change and fix bug in other way

1. revert perivous change
2. add batchTimesWithNoJob Set to record the batch with no job
3. add aggrate method of InputInfo

mwws commented:
@zsxwing Thanks for your suggestion, I have re-implemented this PR according to that. But instead of maintaining state `lastBatchTime` in `JobScheduler`, I would prefer record such info in `JobGenerator`, so we can have central place to resolve inputInfo. 

I according to my manual test the new code works. Could you help review again?

SparkQA commented:
**[Test build #59730 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/59730/consoleFull)** for PR 13259 at commit [`a3d05e0`](https://github.com/apache/spark/commit/a3d05e08c9889bc7a2d4df110430b2889858943f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/59730/
Test PASSed.

zsxwing commented:
@mwws sorry for the delay. Could you add a unit test, please?

wei-mao-intel committed:
add unit test

SparkQA commented:
**[Test build #60095 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60095/consoleFull)** for PR 13259 at commit [`ffd1787`](https://github.com/apache/spark/commit/ffd178779c09b31e6e09e92b1b611d23fe90df4a).

mwws commented:
@zsxwing Yes sir, a unit test has been added.

SparkQA commented:
**[Test build #60095 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/60095/consoleFull)** for PR 13259 at commit [`ffd1787`](https://github.com/apache/spark/commit/ffd178779c09b31e6e09e92b1b611d23fe90df4a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/60095/
Test PASSed.

tdas commented:
@zsxwing Could you take a look at this PR. Would be good to merge this.

mwws commented:
@zsxwing This patch has been pended for a long time, could you help to review?
